# Exercise 4.5

# List of Contents

## Data Consistency Check Procedures
#### 1. Import Libraries and Dataframe
#### 2. Mixed Columns
#### 3. Missing Values
#### 4. Duplicates

## Task Questions
#### Question 2: Describe df_ords
#### Question 3: Mixed-type Data
#### Question 5: Checking for Missing Values
#### Question 6: Addressing Missing Values
#### Question 7: Checking for Duplicate Values
#### Question 8: Address Duplicate Values
#### Question 9: Export Data



#### 1. Import Libraries and Dataframe

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path = r'/Users/tristansavella/Instacart Basket Analysis'

In [3]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', '4.3_orders_products', 'products.csv'), index_col = False)

In [4]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [5]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_passed_since_last_order  
count        3.421083e+06       3.421083e+06                  3.214874e+06  
mean         2.776219e+00       1.345202e+01                  1.111484e+01  
std          2.046829e+00       4.226088e+00                  9.206737e+00  
min          0.000000e+00       0.000000e+00                  0.000000e+00  
25%          1.000000e+00       1.000000e+01                  4.000000e+00  
50%          3.000000e+00       1.300000e+01                  7.000000e+00  
75%          5.000000e+00       1.600000e+01                  1.500000e+01  
max          6.000000e+00       2.300000e+01                  3.000000e+01

#### 2. Mixed Columns

In [6]:
# Create a Dataframe
df_test = pd.DataFrame()

In [7]:
# Create a Mixed-Type Column
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
#testing for mixed data type
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
df_test['mix'] = df_test['mix'].astype('str')

#### 3. Missing Values

In [11]:
# Missing Values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# Create dataframe to show missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
df_prods.shape

(49693, 5)

In [15]:
# Creating subset df for non-missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
df_prods_clean.shape

(49677, 5)

#### 4. Duplicates

In [17]:
# Finding and Addressing Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
# Checking number of rows - there should be 5 less after removing duplicates
df_prods_clean.shape

(49677, 5)

In [20]:
# creating new dataframe without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
# success! there are 5 less
df_prods_clean_no_dups.shape

(49672, 5)

In [22]:
#export df
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked_4.5.csv'))

# Task Questions:
#### Question 2: Describe df_ords

Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

Tip: Keep an eye on min and max values!

In [23]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_passed_since_last_order  
count        3.421083e+06       3.421083e+06                  3.214874e+06  
mean         2.776219e+00       1.345202e+01                  1.111484e+01  
std          2.046829e+00       4.226088e+00                  9.206737e+00  
min          0.000000e+00       0.000000e+00                  0.000000e+00  
25%          1.000000e+00       1.000000e+01                  4.000000e+00  
50%          3.000000e+00       1.300000e+01                  7.000000e+00  
75%          5.000000e+00       1.600000e+01                  1.500000e+01  
max          6.000000e+00       2.300000e+01                  3.000000e+01

I do not see anything odd about the 'df_ords'

#### Question 3: Mixed-type Data

Check for mixed-type data in your df_ords dataframe.

In [24]:
# Question 4: If you find mixed-type data, fix it. The column in question should contain observations of a single data type.
# No mixed data types

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### Question 5: Checking for Missing Values
Run a check for missing values in your df_ords dataframe.

In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [25]:
# Checking to see which columns have missing values

df_ords.isnull().sum()

Unnamed: 0                           0
order_id                             0
user_id                              0
eval_set                             0
order_number                         0
orders_day_of_week                   0
order_hour_of_day                    0
days_passed_since_last_order    206209
dtype: int64

In [26]:
# Viewing the missing values

df_nan = df_ords[df_ords['days_passed_since_last_order'].isnull() == True]

df_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_passed_since_last_order  
0                         2                  8                           NaN  
11                        2                 11                           NaN  
26                        1                 14                           NaN  
39                        6                 11                           NaN  
45                        3                 12                           NaN  
...                     ...                ...                           ...  
3420930                   4                 12                           NaN  
3420934                   3                 18                           NaN  
3421002                   6                 19                           NaN  
3421019                   1                 15                           NaN  
3421069                   3                 11                           NaN  

[206209 rows x 8 columns]

In [27]:
#checking for unique values in 'user_id'

df_ords['user_id'].unique()

array([     1,      2,      3, ..., 206207, 206208, 206209])

There are 206209 entries in "user_id", which is also how many values are "NaN" in 'days_passed_since_last_order'. This means for all of the missing values, this was a customer's first order and therefore, there is no "last order" (also note that the order number is "1" for each of row in which 'days_passed_since_last_order' = NaN).

#### Question 6: Addressing Missing Values
Address the missing values using an appropriate method.

In a markdown cell, explain why you used your method of choice.

No change is necessary. This was a customer's first order and therefore there can be no "number of days since the last order". I would omit these values if I were asked to do an analysis on the average number of days since a customer's previous order, as it would not make sense to include these rows for such an analysis.

#### Question 7: Checking for Duplicate Values
Run a check for duplicate values in your df_ords data.

In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [28]:
df_ords_dups = df_ords[df_ords.duplicated()]
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_passed_since_last_order]
Index: []

There are no duplicates for this dataframe

#### Question 8: Address Duplicate Values

Address the duplicates using an appropriate method.

In a markdown cell, explain why you used your method of choice.

There were no duplicates

#### Question 9: Export Data
Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [94]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked_4.5.csv'))